From 856afa4a358482219dc9a43f79b1e31e12e06e83 Mon Sep 17 00:00:00 2001
From: John Galt <johngaltfirstrun@gmail.com>
Date: Wed, 19 Aug 2026 08:59:56 -0400
Subject: [PATCH 24/39] base: PropImitationHooks: Fix applying spoofing on 17

By using theUnsafe

Better than using hacks on android_art.
[hridaya: Port over from PixelPropsUtils]

Co-authored-by: Ido Ben-Hur <idoybh2@gmail.com>
Change-Id: I86aa4c5b73729c4f005d113ad93b4dde052fe621
Signed-off-by: hridaya <hridaya@pixelos.net>
---
 .../internal/util/PropImitationHooks.java     | 37 ++++++++++++++++++-
 1 file changed, 36 insertions(+), 1 deletion(-)

diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
index 901392220410..b545a87d3c00 100644
--- a/core/java/com/android/internal/util/PropImitationHooks.java
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -52,6 +52,8 @@ import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 
+import sun.misc.Unsafe;
+
 /**
  * @hide
  */
@@ -119,6 +121,29 @@ public class PropImitationHooks {
     private static volatile String sProcessName;
     private static volatile boolean sIsGms, sIsFinsky, sIsPhotos;
 
+    private static final Field OFFSET_FIELD;
+    private static final Unsafe UNSAFE;
+
+    static {
+        Unsafe unsafe = null;
+        Field offsetField = null;
+
+        try {
+            Field field = Unsafe.class.getDeclaredField("theUnsafe");
+            field.setAccessible(true);
+            unsafe = (Unsafe) field.get(null);
+            field.setAccessible(false);
+
+            offsetField = Field.class.getDeclaredField("offset");
+            offsetField.setAccessible(true);
+        } catch (Exception e) {
+            Log.e(TAG, "Unable to initialize Unsafe", e);
+        }
+
+        UNSAFE = unsafe;
+        OFFSET_FIELD = offsetField;
+    }
+
     public static void setProps(Context context) {
         final String packageName = context.getPackageName();
         final String processName = Application.getProcessName();
@@ -166,6 +191,11 @@ public class PropImitationHooks {
     }
 
     private static void setPropValue(String key, String value) {
+        if (UNSAFE == null || OFFSET_FIELD == null) {
+            Log.e(TAG, "Unsafe is unavailable", new IllegalStateException());
+            return;
+        }
+
         try {
             dlog("Setting prop " + key + " to " + value.toString());
             Class clazz = Build.class;
@@ -176,7 +206,12 @@ public class PropImitationHooks {
             Field field = clazz.getDeclaredField(key);
             field.setAccessible(true);
             // Cast the value to int if it's an integer field, otherwise string.
-            field.set(null, field.getType().equals(Integer.TYPE) ? Integer.parseInt(value) : value);
+            long offset = OFFSET_FIELD.getInt(field);
+            if (field.getType().equals(Integer.TYPE)) {
+                UNSAFE.putInt(field.getDeclaringClass(), offset, Integer.parseInt(value));
+            } else {
+                UNSAFE.putObject(field.getDeclaringClass(), offset, value);
+            }
             field.setAccessible(false);
         } catch (Exception e) {
             Log.e(TAG, "Failed to set prop " + key, e);
-- 
2.43.0

